File: Ruby/Euler015.rb
#!/usr/bin/env ruby

def main
  target = 20
  puts combinatoric(target * 2, target)
end

def combinatoric(n, r)
  numer = factorial(n)
  denom = factorial(r) * factorial(n - r)
  return numer / denom
end

def factorial(n)
  p = 1
  while n > 1
    p *= n
    n -= 1
  end
  return p
end

if __FILE__ == $0
  main
end

File: Kotlin/Euler015.kt
package Euler015

fun main() {
  val target = 20L

  println(combinatoric(target * 2, target))
}

fun combinatoric(n: Long, r: Long): Long {
  val numer = factorial(n)
  val denom = factorial(r) * factorial(n - r)
  return Math.round(numer / denom).toLong()
}

fun factorial(n: Long): Double {
  return (n downTo 2).asSequence().map { x -> x.toDouble() }.reduce { a, b -> a * b }
}
